File: source/Controllers/NodeListMixedController.vb

Imports Contensive.BaseClasses
Imports Contensive.Models.Db

Namespace Contensive.AdminNavigator
    Public Class NodeListMixedController
        '
        '====================================================================================================
        ''' <summary>
        ''' list mixed nodes (settings/reports/tools)   
        ''' </summary>
        ''' <param name="cp"></param>
        ''' <param name="env"></param>
        ''' <param name="xEmptyNodeList"></param>
        ''' <param name="TopParentNode"></param>
        ''' <param name="AutoManageAddons"></param>
        ''' <param name="NodeType"></param>
        ''' <param name="xOpenNodeList"></param>
        ''' <param name="AddonNavTypeID"></param>
        ''' <param name="MenuParentNodeID"></param>
        ''' <param name="AdminNavIconTypeSetting"></param>
        ''' <param name="Return_DraggableJS"></param>
        ''' <returns></returns>
        Friend Shared Function getNodeListMixed(cp As CPBaseClass, env As ApplicationEnvironmentModel, xEmptyNodeList As String, TopParentNode As String, AutoManageAddons As Boolean, NodeType As NodeTypeEnum, xOpenNodeList As String, AddonNavTypeID As Integer, MenuParentNodeID As Integer, AdminNavIconTypeSetting As Integer, ByRef Return_DraggableJS As String) As String
            Dim returnNav As String = ""
            Try
                Dim NodeDraggableJS As String
                Dim lastName As String
                Dim sortedNodes As New List(Of KeyValuePair(Of String, SortNodeType))
                Dim SortPtr As Integer
                Dim NodeIDString As String
                Dim Criteria As String
                Dim BlockSubNodes As Boolean
                '
                ' list mixed nodes (settings/reports/tools), includes menu nodes and addons with type='setting' sorted in
                '
                If InStr(1, env.EmptyNodeList & ",", "," & TopParentNode & ",") <> 0 Then
                    env.EmptyNodeList = env.EmptyNodeList
                Else
                    '
                    ' Add addons to node list
                    '
                    NodeIDString = ""
                    Criteria = "(navtypeid=" & AddonNavTypeID & ")"
                    If (AddonNavTypeID = 2) Or (AddonNavTypeID = 3) Or (AddonNavTypeID = 4) Then
                        '
                        ' if setting, report or tool, "admin" is not needed
                        '
                    Else
                        '
                        ' for Manage Addons node, admin is needed for non-developer
                        '
                        If Not cp.User.IsDeveloper() Then
                            Criteria = Criteria & "and(admin<>0)"
                        End If
                    End If
                    Using cs10 As CPCSBaseClass = cp.CSNew
                        If cs10.Open("add-ons", Criteria, "name") Then
                            Do
                                Dim nodeName As String = Trim(cs10.GetText("name"))
                                Dim sortNode As New SortNodeType() With {
                                        .Name = Trim(cs10.GetText("name")),
                                        .addonid = cs10.GetInteger("ID"),
                                        .NavIconTitle = .Name,
                                        .ContentControlID = cs10.GetInteger("ContentControlID"),
                                        .NavIconType = AdminNavIconTypeSetting
                                    }
                                sortedNodes.Add(New KeyValuePair(Of String, SortNodeType)(nodeName, sortNode))
                                SortPtr += 1
                                cs10.GoNext()
                            Loop While cs10.OK
                        End If
                        cs10.Close()
                    End Using
                    '
                    ' Add real navigator nodes to node list
                    '
                    Using cs11 As CPCSBaseClass = cp.CSNew
                        If cs11.OpenSQL(MenuSqlController.getMenuSQL(cp, "parentid=" & MenuParentNodeID, "")) Then
                            Do
                                Dim nodeName As String = Trim(cs11.GetText("name"))
                                Dim navIconTitle As String = cs11.GetText("NavIconTitle")
                                Dim sortNode As New SortNodeType() With {
                                    .Name = Trim(cs11.GetText("name")),
                                    .NavigatorID = cs11.GetInteger("ID"),
                                    .CollectionID = cs11.GetInteger("CollectionID"),
                                    .NewWindow = cs11.GetBoolean("newwindow"),
                                    .contentID = cs11.GetInteger("ContentID"),
                                    .Link = Trim(cs11.GetText("LinkPage")),
                                    .addonid = cs11.GetInteger("AddonID"),
                                    .NavIconType = cs11.GetInteger("NavIconType"),
                                    .NavIconTitle = If(String.IsNullOrEmpty(navIconTitle), nodeName, navIconTitle),
                                    .HelpAddonID = cs11.GetInteger("HelpAddonID"),
                                    .helpCollectionID = cs11.GetInteger("HelpCollectionID"),
                                    .ContentControlID = cs11.GetInteger("ContentControlID"),
                                    .NodeIDString = If(nodeName.ToLowerInvariant() = "all content", NodeIDAllContentList, NodeIDAllContentList)
                                }
                                If (nodeName.ToLowerInvariant() = "all content") Then
                                    sortNode.NodeIDString = NodeIDAllContentList
                                Else
                                    sortNode.NodeIDString = sortNode.NavigatorID.ToString()
                                    sortNode.NavIconTitle = nodeName
                                End If
                                sortedNodes.Add(New KeyValuePair(Of String, SortNodeType)(nodeName, sortNode))
                                SortPtr = SortPtr + 1
                                Call cs11.GoNext()
                            Loop While cs11.OK
                        End If
                        Call cs11.Close()
                    End Using
                    '
                    sortedNodes.Sort(Function(valueA, valueB) valueA.Key.CompareTo(valueB.Key))
                    '
                    If sortedNodes.Count = 0 Then
                        '
                        ' Empty list, add to env.EmptyNodeList
                        '
                        env.EmptyNodeList = env.EmptyNodeList & "," & TopParentNode
                    Else
                        lastName = ""
                        Dim contentNodeList As String = ""
                        For Each kvp In sortedNodes
                            Dim nodeName As String = kvp.Key.ToLowerInvariant()
                            If nodeName <> lastName Then
                                Dim sortedNode As SortNodeType = kvp.Value
                                If (Not sortedNode.contentID.Equals(0)) Then
                                    '
                                    ' -- content nodes last
                                    contentNodeList &= NodeController.GetNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, Nothing, 0, sortedNode.Name, env.EmptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.OpenNodeList, sortedNode.NodeIDString, NodeDraggableJS, "")
                                ElseIf (sortedNode.addonid.Equals(0)) Then
                                    '
                                    ' -- hard-coded nodes
                                    returnNav &= NodeController.GetNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, Nothing, 0, sortedNode.Name, env.EmptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.OpenNodeList, sortedNode.NodeIDString, NodeDraggableJS, "")
                                Else
                                    '
                                    ' -- addon nodes next
                                    Dim addon As AddonModel = DbBaseModel.create(Of AddonModel)(cp, sortedNode.addonid)
                                    returnNav &= NodeController.GetNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, addon, 0, sortedNode.Name, env.EmptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.OpenNodeList, sortedNode.NodeIDString, NodeDraggableJS, "")
                                End If
                                Return_DraggableJS &= NodeDraggableJS
                                lastName = nodeName
                            End If
                        Next
                        returnNav &= contentNodeList
                    End If
                End If
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
                Throw
            End Try
            Return returnNav
        End Function
    End Class
End Namespace

File: source/Controllers/NodeListMixedController.cs
using System;
using System.Collections.Generic;
using Contensive.BaseClasses;
using Contensive.Models.Db;

namespace Contensive.AdminNavigator {
    public sealed class NodeListMixedController {
        // 
        // ====================================================================================================
        /// <summary>
        /// list mixed nodes (settings/reports/tools)   
        /// </summary>
        /// <param name="cp"></param>
        /// <param name="env"></param>
        /// <param name="xEmptyNodeList"></param>
        /// <param name="TopParentNode"></param>
        /// <param name="AutoManageAddons"></param>
        /// <param name="NodeType"></param>
        /// <param name="xOpenNodeList"></param>
        /// <param name="AddonNavTypeID"></param>
        /// <param name="MenuParentNodeID"></param>
        /// <param name="AdminNavIconTypeSetting"></param>
        /// <param name="Return_DraggableJS"></param>
        /// <returns></returns>
        internal static string getNodeListMixed(CPBaseClass cp, ApplicationEnvironmentModel env, string xEmptyNodeList, string TopParentNode, bool AutoManageAddons, common.NodeTypeEnum NodeType, string xOpenNodeList, int AddonNavTypeID, int MenuParentNodeID, int AdminNavIconTypeSetting, ref string Return_DraggableJS) {
            string returnNav = "";
            try {
                string lastName;
                var sortedNodes = new List<KeyValuePair<string, common.SortNodeType>>();
                var SortPtr = default(int);
                string NodeIDString;
                string Criteria;
                var BlockSubNodes = default(bool);
                // 
                // list mixed nodes (settings/reports/tools), includes menu nodes and addons with type='setting' sorted in
                // 
                if ((env.emptyNodeList + ",").IndexOf("," + TopParentNode + ",", StringComparison.Ordinal) >= 0) {
                    // 
                } else {
                    // 
                    // Add addons to node list
                    // 
                    NodeIDString = "";
                    Criteria = "(navtypeid=" + AddonNavTypeID + ")";
                    if (AddonNavTypeID == 2 | AddonNavTypeID == 3 | AddonNavTypeID == 4) {
                        // 
                        // if setting, report or tool, "admin" is not needed
                        // 
                        // 
                        // for Manage Addons node, admin is needed for non-developer
                        // 
                    } else if (!cp.User.IsDeveloper) {
                        Criteria += "and(admin<>0)";
                    }
                    using (var cs10 = cp.CSNew()) {
                        if (cs10.Open("add-ons", Criteria, "name")) {
                            do {
                                string nodeName = cs10.GetText("name").Trim();
                                var sortNode = new common.SortNodeType() {
                                    Name = cs10.GetText("name").Trim(),
                                    addonid = cs10.GetInteger("ID"),
                                    NavIconTitle = cs10.GetText("name").Trim(),
                                    ContentControlID = cs10.GetInteger("ContentControlID"),
                                    NavIconType = AdminNavIconTypeSetting
                                };
                                sortedNodes.Add(new KeyValuePair<string, common.SortNodeType>(nodeName, sortNode));
                                SortPtr += 1;
                                cs10.GoNext();
                            }
                            while (cs10.OK());
                        }
                        cs10.Close();
                    }
                    // 
                    // Add real navigator nodes to node list
                    // 
                    using (var cs11 = cp.CSNew()) {
                        if (cs11.OpenSQL(MenuSqlController.getMenuSQL(cp, "parentid=" + MenuParentNodeID, ""))) {
                            do {
                                string nodeName = cs11.GetText("name").Trim();
                                string navIconTitle = cs11.GetText("NavIconTitle");
                                var sortNode = new common.SortNodeType() {
                                    Name = cs11.GetText("name").Trim(),
                                    NavigatorID = cs11.GetInteger("ID"),
                                    CollectionID = cs11.GetInteger("CollectionID"),
                                    NewWindow = cs11.GetBoolean("newwindow"),
                                    contentID = cs11.GetInteger("ContentID"),
                                    Link = cs11.GetText("LinkPage").Trim(),
                                    addonid = cs11.GetInteger("AddonID"),
                                    NavIconType = cs11.GetInteger("NavIconType"),
                                    NavIconTitle = string.IsNullOrEmpty(navIconTitle) ? nodeName : navIconTitle,
                                    HelpAddonID = cs11.GetInteger("HelpAddonID"),
                                    helpCollectionID = cs11.GetInteger("HelpCollectionID"),
                                    ContentControlID = cs11.GetInteger("ContentControlID"),
                                    NodeIDString = nodeName.ToLowerInvariant() == "all content" ? common.NodeIDAllContentList : common.NodeIDAllContentList
                                };
                                if (nodeName.ToLowerInvariant() == "all content") {
                                    sortNode.NodeIDString = common.NodeIDAllContentList;
                                } else {
                                    sortNode.NodeIDString = sortNode.NavigatorID.ToString();
                                    sortNode.NavIconTitle = nodeName;
                                }
                                sortedNodes.Add(new KeyValuePair<string, common.SortNodeType>(nodeName, sortNode));
                                SortPtr += 1;
                                cs11.GoNext();
                            }
                            while (cs11.OK());
                        }
                        cs11.Close();
                    }
                    // 
                    sortedNodes.Sort((valueA, valueB) => valueA.Key.CompareTo(valueB.Key));
                    // 
                    if (sortedNodes.Count == 0) {
                        // 
                        // Empty list, add to env.EmptyNodeList
                        // 
                        env.emptyNodeList += "," + TopParentNode;
                    } else {
                        lastName = "";
                        string contentNodeList = "";
                        foreach (var kvp in sortedNodes) {
                            string nodeName = kvp.Key.ToLowerInvariant();
                            if ((nodeName ?? "") != (lastName ?? "")) {
                                var sortedNode = kvp.Value;
                                string NodeDraggableJS = "";
                                if (!sortedNode.contentID.Equals(0)) {
                                    // 
                                    // -- content nodes last
                                    contentNodeList += NodeController.getNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, null, 0, sortedNode.Name, env.emptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.openNodeList, sortedNode.NodeIDString, ref NodeDraggableJS, "");
                                } else if (sortedNode.addonid.Equals(0)) {
                                    // 
                                    // -- hard-coded nodes
                                    returnNav += NodeController.getNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, null, 0, sortedNode.Name, env.emptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.openNodeList, sortedNode.NodeIDString, ref NodeDraggableJS, "");
                                } else {
                                    // 
                                    // -- addon nodes next
                                    var addon = DbBaseModel.create<AddonModel>(cp, sortedNode.addonid);
                                    returnNav += NodeController.getNode(cp, env, sortedNode.CollectionID, sortedNode.ContentControlID, sortedNode.helpCollectionID, sortedNode.HelpAddonID, sortedNode.contentID, sortedNode.Link, addon, 0, sortedNode.Name, env.emptyNodeList, sortedNode.NavigatorID, sortedNode.NavIconType, cp.Utils.EncodeHTML(sortedNode.NavIconTitle), AutoManageAddons, NodeType, sortedNode.NewWindow, BlockSubNodes, env.openNodeList, sortedNode.NodeIDString, ref NodeDraggableJS, "");
                                }
                                Return_DraggableJS += NodeDraggableJS;
                                lastName = nodeName;
                            }
                        }
                        returnNav += contentNodeList;
                    }
                }
            } catch (Exception ex) {
                cp.Site.ErrorReport(ex);
                throw;
            }
            return returnNav;
        }
    }
}